src/misc/VarNames.jl

```diff
@@ -0,0 +1,576 @@
+import MacroTools as MT
+import Printf
+
+@assert VarName === Union{Symbol, AbstractString, Char}
+const VarNames = Union{
+    VarName,
+    AbstractArray{<:VarName},
+    Pair{<:VarName},
+}
+
+req(cond, msg) = cond || throw(ArgumentError(msg))
+# The macro version will only evaluate `msg` if needed
+macro req(cond, msg) :($cond || throw(ArgumentError($msg))) end
+
+@doc raw"""
+    variable_names(a...) -> Vector{Symbol}
+    variable_names(a::Tuple) -> Vector{Symbol}
+
+Create proper variable names from `a`.
+Each argument can be either an Array of `VarName`s, `s::VarName => (iter...)`, or `s::VarName => iter`.
+The `iter` version is equivalent to `["s[$i]" for i in iter]`; the `(iter...)` version is the same for multiple indices.
+
+By default `:x => axes` and `"x" => axes` create variables like `x[1,1]`.
+By using `"x#" => axes` instead, `x[1,1]` becomes `x11`, `x[10,10]` becomes `x10_10`, and `x[-1]` becomes `xm1`.
+If one does not want to replace the chars "-./" by "mpq", use '@' instead of '#'.
+Using multiple '#' or multiple '@' as in "x#y#", one gets variables like `x1y1`.
```

Comment:
If we don't support these special substitutions, then we can also get rid of `@`. I really would strongly prefer to keep this as simple as possible.